lib/upipe-modules/upipe_vanc_decoder.c

```diff
@@ -0,0 +1,325 @@
+        upipe_dbg_va(upipe, "Packet too small (%d)", end);
+        goto ret;
+    }
+
+    while (end >= 9) {
+        if (r[0] & 0xfc) { /* 000000 */
+            goto ret;
+        }
+
+        bool c_not_y = r[0] & 0x2;
+
+        unsigned line = ((r[0] & 1) << 10) | (r[1] << 2) | (r[2] >> 6);
+        unsigned offset = ((r[2] & 0x3f) << 6) | (r[3] >> 2);
+        uint16_t did = ((r[3] & 0x3) << 8) | r[4];
+        uint16_t sdid = (r[5] << 2) | (r[6] >> 6);
+        uint16_t dc = ((r[6] & 0x3f) << 4) | (r[7] >> 4);
+
+        uint8_t cache = r[7] & 0xf;
+        unsigned cached_bits = 4;
+        r += 8;
+        end -= 8;
+
+        size_t bits_left = 10 * ((dc & 0xff) + 1 /* checksum */) - cached_bits;
+        if (((bits_left + 7) / 8) > end) {
+            upipe_dbg_va(upipe, "Invalid DC %u, packet size %d", dc & 0xff, end);
+            goto ret;
+        }
+
+        end -= (bits_left + 7) / 8;
+
+        if (line == 0) {
+            upipe_dbg(upipe, "Invalid line number 0");
+            r += (bits_left + 7) / 8;
+            continue;
+        }
+
+        struct uref *pic = uref_pic_alloc(vancd->uref_mgr, vancd->ubuf_mgr,
+                S291_HEADER_SIZE + (dc & 0xff) + 1, 1);
+        if (unlikely(!pic)) {
+            upipe_throw_fatal(upipe, UBASE_ERR_ALLOC);
+            goto ret;
+        }
+
+        uint8_t *vanc_buf;
+        if (unlikely(!ubase_check(uref_pic_plane_write(pic, "x10",
+                            0, 0, -1, -1, &vanc_buf)))) {
+            uref_free(pic);
+        }
+        uint16_t *data = (uint16_t*)vanc_buf;
+
+        data[0] = S291_ADF1;
+        data[1] = S291_ADF2;
+        data[2] = S291_ADF3;
+        data[3] = did;
+        data[4] = sdid;
+        data[5] = dc;
+
+        for (int i = 0; i < (dc & 0xff) + 1; i++) {
+            if (cached_bits == 0) {
+                cache = *r++;
+                cached_bits = 8;
+            }
+
+            unsigned needed_bits = 10 - cached_bits;
+            uint16_t val = cache << needed_bits;
+
+            cache = *r++;
+            cached_bits = 8 - needed_bits;
+
+            val |= cache >> cached_bits;
+
+            cache &= (1 << cached_bits) - 1;
+
+            data[S291_HEADER_SIZE+i] = val;
+        }
+
+        if (cached_bits && cache != (1 << cached_bits) - 1) {
+            upipe_dbg(upipe, "Invalid byte align, skipping");
+            uref_pic_plane_unmap(pic, "x10", 0, 0, -1, -1);
+            uref_free(pic);
+            continue;
+        }
+
+        if (!s291_check_cs(data)) {
+            upipe_dbg(upipe, "Invalid checksum, skipping");
+            uref_pic_plane_unmap(pic, "x10", 0, 0, -1, -1);
+            uref_free(pic);
+            continue;
+        }
+
+        upipe_dbg_va(upipe, "y=%d line %d off %d | DID 0x%.2x SDID 0x%.2x DC 0x%.2x",
+                c_not_y, line, offset,
+                s291_get_did(data), s291_get_sdid(data), s291_get_dc(data));
+
+        if (c_not_y)
+            offset = 2 * offset + 1;
```

Comment:
Wrong as C comes before Y and it's interleaved so we need another bit since we don't know if this is SD or HD